host/class/uvc/usb_host_uvc_2/uvc_host.c

```diff
@@ -0,0 +1,816 @@
+        usb_transfer_t *this_transfer = uvc_stream->constant.xfers[i];
+        this_transfer->device_handle = uvc_stream->constant.dev_hdl;
+        this_transfer->context = uvc_stream;
+        this_transfer->timeout_ms = 1000;
+        this_transfer->bEndpointAddress = ep_desc->bEndpointAddress;
+
+        if (is_isoc) {
+            this_transfer->callback = isoc_transfer_callback;
+            this_transfer->num_bytes = num_isoc_packets * max_packet_size;
+            for (unsigned j = 0; j < num_isoc_packets; j++) {
+                this_transfer->isoc_packet_desc[j].num_bytes = max_packet_size;
+            }
+        } else {
+            this_transfer->callback = bulk_transfer_callback;
+            this_transfer->num_bytes = transfer_size;
+        }
+    }
+    return ESP_OK;
+
+err:
+    uvc_transfers_free(uvc_stream);
+    return ret;
+}
+
+/**
+ * @brief Helper function that releases resources claimed by UVC device
+ *
+ * Close underlying USB device, free device driver memory
+ *
+ * @note All interfaces claimed by this device must be release before calling this function
+ * @param uvc_stream UVC stream handle to be removed
+ */
+static void uvc_device_remove(uvc_stream_t *uvc_stream)
+{
+    assert(uvc_stream);
+    uvc_transfers_free(uvc_stream);
+    uvc_frame_free(uvc_stream);
+    // We don't check the error code of usb_host_device_close, as the close might fail, if someone else is still using the device (not all interfaces are released)
+    usb_host_device_close(p_uvc_host_driver->usb_client_hdl, uvc_stream->constant.dev_hdl); // Gracefully continue on error
+    free(uvc_stream);
+}
+
+/**
+ * @brief Open USB device with requested VID/PID
+ *
+ * This function has two regular return paths:
+ * 1. USB device with matching VID/PID is already opened by this driver: allocate new UVC device on top of the already opened USB device.
+ * 2. USB device with matching VID/PID is NOT opened by this driver yet: poll USB connected devices until it is found.
+ *
+ * @note This function will block for timeout_ticks, if the device is not enumerated at the moment of calling this function.
+ * @param[in]  vid           Vendor ID
+ * @param[in]  pid           Product ID
+ * @param[in]  timeout_ticks Connection timeout in FreeRTOS ticks
+ * @param[out] dev           UVC device
+ * @return
+ *     - ESP_OK: Success - device opened
+ *     - ESP_ERR_NOT_FOUND: Device not found in given timeout
+ */
+static esp_err_t uvc_find_and_open_usb_device(uint16_t vid, uint16_t pid, TickType_t timeout_ticks, uvc_stream_t **dev)
+{
+    assert(p_uvc_host_driver);
+    assert(dev);
+
+    *dev = calloc(1, sizeof(uvc_stream_t));
+    if (*dev == NULL) {
+        return ESP_ERR_NO_MEM;
+    }
+
+    // First, check list of already opened UVC devices
+    ESP_LOGD(TAG, "Checking list of opened USB devices");
+    uvc_stream_t *uvc_stream;
+    SLIST_FOREACH(uvc_stream, &p_uvc_host_driver->uvc_stream_list, list_entry) {
```

Comment:
Maybe, when the `vid/pid == 0`, we don't need to iterate the list, but to return the first member of the list. 
Seems, that otherwise the logic is bit complicated with iteration for both cases.

Reply:
There are also cases when
1. pid == 0 && vid != 0
2. pid != 0 && vid == 0

The loop correctly handles all cases, I don't want to add 'exceptions' for the algorithm

Reply:
Yeah, but are these cases are real use-cases?
I think the most common scenarios are: to use the special device with specific vid/pid or to use any device with any vid/pid. 
Situation, when we need to use any device by the specific pid just seems very specific for me.